### PR TITLE
[Query Expression]Restrict result type being resolved as `map` without the query-construct-type

### LINKED ISSUE #37256: Create Variable Code Action is not accurate for query expression
**Description:**
Consider the following scenario
Peek 2022-08-04 13-45

**Steps to reproduce:**
```Ballerina
public function main() {
    xml x1 = xml `<book>The Lost World</book>`;

    string[] xmlResult = from xml element in x1 select element.toBalString();
}

```
**Affected Versions:**
2201.2.x

### LINKED ISSUE #37450: Incorrect diagnostic property in query expression
**Description:**
Consider the diagnostic `variable assignment is required` at cursor position.

Screenshot 2022-08-19 at 09 29 30

In the diagnostic properties which is used to derive the expected type, we get the `bType` incorrectly as `xml<string>` where it should be `string[]`.

Screenshot 2022-08-19 at 09 28 14

For a different case like `x1.toBalString()`, we get the type correctly as `string` in the diagnostic properties.

**Steps to reproduce:**
```
public function main() {
    xml x1 = xml `<book>The Lost World</book>`;

    from xml element in x1 select element.toBalString();
}
```

**Affected Versions:**
2201.2.0 (Swan Lake Update 2)

### LINKED ISSUE #37670: [Bug]: When inferring type for query-exp, collections of types other than query construct types should throw an error
### Description

From the spec https://ballerina.io/spec/lang/master/#section_6.35:
> If there is no contextually expected type, then the basic type of the value constructed is the basic type of expression following in; it is a compile-time error if the static type of this expression is not a subtype of one of the basic types that a query expression can construct.

```ballerina
import ballerina/io;
public function main() {
    xml x = xml ``;
    var res = from xml i in x // we should get an error here incompatible types: 'xml' cannot be constrained with 'int'
    select 1;

    io:println(res is int[]); //error incompatible types: 'xml<int>' will not be matched to 'int[]'
}
```

```ballerina
public function main() {
    var res = from string i in "x"
    select 1; // atm, no errors. should be an error

    string x = res;
}
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.2.0

### OS, DB, other environment details and versions

_No response_

### Labels

CompilerFE

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #41129: [Bug]: Stream toString representation is incorrect
### Description

```ballerina 
import ballerina/io;

public function main() {
    stream<int> streamResult = [1, 2, 3].toStream();
    io:println(streamResult);
}
```

Screenshot 2023-07-31 at 12 20 24

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.7.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #37450
Fixes #37670
Fixes #41129
Fixes #37256

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by pcnfernando** [comment]
Keeping PR as draft until https://github.com/ballerina-platform/ballerina-spec/issues/1252 is finalized

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java [hunk 1]
```diff
@@ -570,26 +574,40 @@ private BType getResolvedType(BType initType, BType expType, boolean isReadonly,
         return initType;
     }
 
-    private BType getNonContextualQueryType(BType staticType, BType basicType) {
-        BType resultType;
+    private BType getNonContextualQueryType(BType constraintType, BType basicType,  Location pos) {
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java:577]:
```suggestion
    private BType getNonContextualQueryType(BType constraintType, BType basicType, Location pos) {
```

**Review by NipunaRanasinghe — APPROVED** [review]
Debugger changes LGTM

**Review by malinthar — CHANGES_REQUESTED** [review]
Can we add a LS code action test case for the following as well

```Ballerina
public function main() {
    xml x1 = xml `<book>The Lost World</book>`;
    from xml element in x1 select element.toBalString();
}

```

**Comment by pcnfernando** [comment]
> Can we add a LS code action test case for the following as well
> 
> ```ballerina
> public function main() {
>     xml x1 = xml `<book>The Lost World</book>`;
>     from xml element in x1 select element.toBalString();
> }
> ```

Inferring would result in an error for this specific scenario. Therefore, no code action. Can you point me to a such a codeaction test I can refer

**Comment by malinthar** [comment]
> > Can we add a LS code action test case for the following as well
> > ```ballerina
> > public function main() {
> >     xml x1 = xml `<book>The Lost World</book>`;
> >     from xml element in x1 select element.toBalString();
> > }
> > ```
> 
> Inferring would result in an error for this specific scenario. Therefore, no code action. Can you point me to a such a codeaction test I can refer

+1. We can add a negative test case in that case

**Review by malinthar — APPROVED** [review]

**Pull request merged**